### PR TITLE
rgw: fix a bug the owner should delete objects in the container

### PR DESCRIPTION
**Pull request by Jing-Scott**
The owner should has right to delete the object in his container by default

Fixes: http://tracker.ceph.com/issues/18517
Signed-off-by: Jing Wenjun <jingwenjun@cmss.chinamobile.com>

### CODE REVIEW

**Comment by Jing-Scott** [comment]
@cbodley Hi,
Can you help to review? Thanks

**Comment by Jing-Scott** [comment]
@cbodley Hi,
Can you help to review? Thanks.

**Comment by cbodley** [comment]
@Jing-Scott I'm investigating this with help from @rzarzynski. I don't think this is the right fix, as it changes the semantics of s3 operations as well. It looks like we're not handling these swift acl operations correctly - I'll update http://tracker.ceph.com/issues/18517 with more info.

File: src/rgw/rgw_op.cc [hunk 1]
```diff
@@ -3659,7 +3659,7 @@ int RGWDeleteObj::handle_slo_manifest(bufferlist& bl)
 
 int RGWDeleteObj::verify_permission()
 {
-  if (!verify_bucket_permission(s, RGW_PERM_WRITE)) {
+  if (!verify_bucket_permission(s, RGW_PERM_WRITE) && s->user->user_id.compare(s->bucket_owner.get_id()) != 0) {
```
Comment (cbodley) [inline src/rgw/rgw_op.cc]: this change in `RGWDeleteObj` is not actually needed - we're only rejecting the HEAD request with '403: Forbidden'. `verify_bucket_permission()` here is correctly detecting the owner's `RGW_PERM_WRITE` permission

**Comment by Jing-Scott** [comment]
@cbodley Yeah. Thanks for your detailed analysis. What you pointed out in the tracker is meaningful for me.

**Comment by Jing-Scott** [comment]
@cbodley Hi,
I've modified the commit according to your suggestion. This change will not affect the s3 operation, I have explained in my commit message. Thanks

**Comment by Jing-Scott** [comment]
@cbodley  Sorry! I reset the last commit to the original. what I commited was not correct.

**Comment by Jing-Scott** [comment]
@cbodley Hi,
The first approach what you put forwad on the tracker is the most ideal way to handle currently.

I've updated the commit. Pls check it.

The container created by swift with the` SWIFT_PERM_ADMIN | SWIFT_PERM_READ` as the container owner's permission, which can resolve the problem mentioned in the tracker and I've tested it already.

**Comment by Jing-Scott** [comment]
jenkins test this please

**Comment by cbodley** [comment]
looks good! i'll run this through teuthology for validation. in the meantime, can you please squash your changes into a single commit?

**Comment by Jing-Scott** [comment]
jenkins test this please

**Comment by Jing-Scott** [comment]
jenkins test this please

**Comment by Jing-Scott** [comment]
@cbodley 
OK. I have already squashed them. Thanks a lot!

**Comment by Jing-Scott** [comment]
jenkins test this please

**Comment by Jing-Scott** [comment]
@cbodley 
hmm, i have thought that this solution maybe not good enough.
cuz it will [show the container's owner container read acl](https://github.com/ceph/ceph/blob/master/src/rgw/rgw_acl_swift.cc#L200) by default.

**Comment by cbodley** [comment]
> @cbodley
> hmm, i have thought that this solution maybe not good enough.
> cuz it will show the container's owner container read acl by default.

ah, you're right. i had only looked at the `Read ACL:` entry right after bucket creation, and it didn't show up there. but on bucket creation, we're calling through `get_bucket_policy_from_attr()` to `RGWAccessControlPolicy::create_default()`, which only grants the `RGW_PERM_FULL_CONTROL` permissions

we don't call `RGWAccessControlPolicy_SWIFT::create()` until the `put_bucket_metadata` request that sets the acls. after that, the owner does always show up in the `Read ACL` :(

**Comment by Jing-Scott** [comment]
@cbodley : this modification maybe not the most ideal way, however, which is one of the effective ways to handle in the current case.

**Pull request closed without merging**